M2/Macaulay2/packages/SparseResultants.m2

```diff
@@ -462,6 +463,210 @@ genericMultihomogeneousPolynomial (VisibleList,VisibleList) := o -> (k,d) -> (
     sum(gens A,e -> (b = last baseName e; e * product(n,i -> (mm_i)_(b_i))))
 );
 
+-- Hyperdeterminants --
+
+MultidimensionalMatrix = new Type of HashTable;
+
+MultidimensionalMatrix#{Standard,AfterPrint} = MultidimensionalMatrix#{Standard,AfterNoPrint} = M -> (
+    << endl << concatenate(interpreterDepth:"o") << lineNumber << " : " << "Multidimensional matrix of format " << M#"format" << " over " << M#"ring" << endl;
+);
+
+net MultidimensionalMatrix := M -> net(M#"list");
+
+multidimensionalMatrix = method(TypicalValue => MultidimensionalMatrix);
+
+multidimensionalMatrix (List) := (L) -> (
+    X := makeRing L;
+    F := makeMultilinearForm(L,X);
+    L' := makeList(F,X);
+    new MultidimensionalMatrix from {
+        "format" => apply(X,i -> #i),
+        "list" => L',
+        "multilinearForm" => F,
+        "varsMultilinearForm" => X,
+        "ring" => coefficientRing ring F  
+    }
+);
+
+Thing * MultidimensionalMatrix := (a,M) -> (
+    applyToEntries(M,m -> a*m)
+);
+
+RingElement * MultidimensionalMatrix := (a,M) -> (
+    applyToEntries(M,m -> a*m)
+);
+
+- MultidimensionalMatrix := (M) -> (-1) * M;
+
+MultidimensionalMatrix - MultidimensionalMatrix := (M,N) -> (
+   if M#"format" != N#"format" then error "expected matrix of the same format";
+   multidimensionalMatrix(M#"list" - N#"list")
+);
+
++ MultidimensionalMatrix := (M) -> M;
+
+MultidimensionalMatrix + MultidimensionalMatrix := (M,N) -> (
+   if M#"format" != N#"format" then error "expected matrix of the same format";
+   multidimensionalMatrix(M#"list" + N#"list")
+);
+
+MultidimensionalMatrix == MultidimensionalMatrix := (M,N) -> (
+   if M#"format" != N#"format" then return false;
+   if M#"ring" =!= N#"ring" then return false;
+   M#"multilinearForm" == N#"multilinearForm"
+);
+
+transposition = method();
+transposition (MultidimensionalMatrix,List) := (M,l) -> (
+    n := M#"format";
+    try assert(ring matrix {l} === ZZ and sort l == toList(0 .. #n-1)) else error("expected a permutation of the set "|toString toList(0 .. #n-1));
+    K := M#"ring";
+    X := M#"varsMultilinearForm"; 
+    n' := for i to #n-1 list n_(l_i);
+    X' := for i to #n-1 list X_(l_i);
+    R := K[flatten X'];
+    X' = apply(X',X0->apply(X0,u -> sub(u,R)));
+    multidimensionalMatrix makeList(sub(M#"multilinearForm",R),X')
+);
+
+makeRing = method();
+makeRing (List) := (L) -> (
+    if #L == 0 then error "expected a nonempty list";
+    n := {#L};
+    L' := first L;
+    while instance(L',List) do (
+        if #L' == 0 then error "expected nonempty lists";
+        n = prepend(#L',n);
+        L' = first L';
+    );
+    for i to #n-2 do L = flatten L;
+    K := ring matrix{L};
+    gensRing0(K,n)
+);
+
+makeMultilinearForm = method();
+makeMultilinearForm (List,List) := (L,X) -> (
+    if not instance(first L,List) then return ((matrix{last X})*transpose(matrix{L}))_(0,0);
+    X' := X_{0 .. #X-2};
```

Comment:
It's better to use `take` here, for speed.

Reply:
Done.